### ISSUE TITLE
linux-msft-wsl-5.10.93.2 kernel fails to compile

### ISSUE THREAD

**Issue by johnvranos** [microsoft/WSL#8045]
### Version

Microsoft Windows [Version 10.0.22000.493]

### WSL Version

- [X] WSL 2
- [ ] WSL 1

### Kernel Version

5.10.60.1-microsoft-standard-WSL2

### Distro Version

Ubuntu 21.10, installed in WSL 2, with Distrod (https://github.com/nullpo-head/wsl-distrod).

### Repro Steps

The steps described in https://github.com/microsoft/WSL2-Linux-Kernel:

**Build Instructions**

Instructions for building an x86_64 WSL2 kernel with an Ubuntu distribution are as follows:

Install the build dependencies:
$ sudo apt install build-essential flex bison dwarves libssl-dev libelf-dev

Build the kernel using the WSL2 kernel configuration:
$ make KCONFIG_CONFIG=Microsoft/config-wsl

### Expected Behavior

The kernel compiles successfully.

### Actual Behavior

The kernel fails to compile with error:

make[2]: *** No rule to make target 'net/netfilter/xt_HL.o', needed by 'net/netfilter/built-in.a'. Stop.
make[1]: *** [scripts/Makefile.build:497: net/netfilter] Error 2
make: *** [Makefile:1822: net] Error 2

### Diagnostic Logs

![s](https://user-images.githubusercontent.com/51235667/153629719-2f4b787a-988a-4682-ac9d-561fbda98a0c.png)

The image shows what is the error message.

**Comment by benhillis** [comment]
@tyhicks - is something missing from the readme?

**Comment by tyhicks** [comment]
I don't think so as I'm having no problems on Ubuntu 20.04 LTS.

@johnvranos are you patching the kernel or modifying the config?

**Comment by johnvranos** [comment]
No, I am not modifying the configuration.

I run:

tar -xf WSL2-Linux-Kernel-linux-msft-wsl-5.10.93.2.tar.gz

and then I run:

make KCONFIG_CONFIG=Microsoft/config-wsl

Then it asks one question:

KCSAN: dynamic data race detector (KCSAN) [N/y/?] (NEW)

where I press N

Then it continues with the compilation, and after some time the compilation produces the error.

**Comment by johnvranos** [comment]
I am trying to compile in Ubuntu 21.10, NOT in Ubuntu 20.04.

Now I am trying to compile the previous version, 5.10.81.1.

I will let you know if it works.

**Comment by tyhicks** [comment]
Could you run these commands and paste their results?

```
$ sha1sum net/netfilter/Makefile
$ grep NETFILTER_XT_TARGET_HL Microsoft/config-wsl
```

**Comment by johnvranos** [comment]
> I am trying to compile in Ubuntu 21.10, NOT in Ubuntu 20.04.
> 
> Now I am trying to compile the previous version of the kernel, 5.10.81.1.
> 
> I will let you know if it works.

The previous version completed successfully. See the image attached.
![s](https://user-images.githubusercontent.com/51235667/153651711-bdaa74ac-ab07-4804-a3ec-4e62d4de624e.png)

**Comment by johnvranos** [comment]
> 

john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$ sha1sum net/netfilter/Makefile
65a2b1bc090fa4924bf82186e2ebd3faafc006ef  net/netfilter/Makefile
john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$
john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$
john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$
john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$ grep NETFILTER_XT_TARGET_HL Microsoft/config-wsl
CONFIG_NETFILTER_XT_TARGET_HL=y
john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$

**Comment by johnvranos** [comment]
> Could you run these commands and paste their results?
> 
> ```
> $ sha1sum net/netfilter/Makefile
> $ grep NETFILTER_XT_TARGET_HL Microsoft/config-wsl
> ```

john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$ sha1sum net/netfilter/Makefile
65a2b1bc090fa4924bf82186e2ebd3faafc006ef  net/netfilter/Makefile
john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$
john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$
john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$
john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$ grep NETFILTER_XT_TARGET_HL Microsoft/config-wsl
CONFIG_NETFILTER_XT_TARGET_HL=y
john@DESKTOP-3D1NDUV:/mnt/c/Users/John/Downloads/test$

**Comment by johnvranos** [comment]
It is my first time I compiled the Linux kernel, do you know what is the name of the WSL 2 kernel image?

I can not find any file named vmlinuz, or bzimage in the source code directories.

Thank you a lot.

**Comment by tyhicks** [comment]
I'm a bit stumped. I don't have any problem building linux-msft-wsl-5.10.93.2, using the same commands that you listed in this issue, inside of a 21.10 LXD container. Everything works fine here.

**Comment by tyhicks** [comment]
> I can not find any file named vmlinuz, or bzimage, in the source code directories.

You should have a vmlinux file sitting in the top-level source directory.

**Comment by johnvranos** [comment]
> I'm a bit stumped. I don't have any problem building linux-msft-wsl-5.10.93.2, using the same commands that you listed in this issue, inside of a 21.10 LXD container. Everything works fine here.

I have Ubuntu 21.10 and Ubuntu MATE 21.10 as two virtual machines in Hyper-V, I will try to compile the latest WSL 2 kernel in the first one.

**Comment by johnvranos** [comment]
The official Ubuntu 21.10 release I have in Hyper-V, is of low quality, I get dependency errors when I try to do:

sudo apt install build-essential flex bison dwarves libssl-dev libelf-dev

I will make more tests tomorrow.

**Comment by johnvranos** [comment]
Well this is what I did today.

My OS is Windows 11 Pro.


1. **I installed Ubuntu-20.04 instance in WSL 2, with wsl --install -d Ubuntu-20.04.**

2. I installed Ubuntu 21.10 and 20.04 with Distrod.

https://github.com/nullpo-head/wsl-distrod

Distrod downloads and uses Linux distribution containers from linuxcontainers.org.

https://linuxcontainers.org

3. I have various Linux virtual machines in Hyper-V. One of them is a Manjaro MATE virtual machine.

Manjaro is a rolling-release distribution, is based on Arch Linux, and has the latest available software in all categories.

**All of the above have the latest updates installed.**


**In all WSL 2 distributions (cases 1. and 2.), I tried to compile WSL 2 kernel versions 5.10.81.1 and 5.10.93.2 (this is the most recent version).**

**In all cases, compilation failed (including Ubuntu-20.04 from Microsoft Store), for both versions of the kernel.**

**In the Manjaro MATE virtual machine (case 3.), I tried compiling the latest version of the kernel (version 5.10.93.2), and it succeeded.**

Manjaro success:

![manjaro](https://user-images.githubusercontent.com/51235667/153724073-f6cac5d7-9a4a-4e55-852f-d19760c6635e.png)



Ubuntu failures:

![s1](https://user-images.githubusercontent.com/51235667/153724569-cc1fb8e6-4ba9-4319-8db1-3627a5074a45.png)

![s2](https://user-images.githubusercontent.com/51235667/153724570-a9a5136d-9947-40c5-acd2-2766fac7fdaa.png)

![s3](https://user-images.githubusercontent.com/51235667/153724571-3fd0e15f-f261-48e8-a4d4-2bfbcbcc25e4.png)

![s4](https://user-images.githubusercontent.com/51235667/153724575-8dffd9fc-22fe-4380-ad48-ee7515988d2f.png)

![s5](https://user-images.githubusercontent.com/51235667/153724577-307b8f8f-586c-4456-8e59-cf5c55c26c6c.png)

I got the new compiled kernel I wanted, from the Manjaro virtual machine.

I do not know why the compilation fails in Ubuntu 20.04 and 21.10 WSL 2 instances.

**Comment by onomatopellan** [comment]
@johnvranos You are compiling the kernel in a `/mnt/*` folder which is not supported. Always extract and compile the kernel in a `/home/user` linux folder.

**Comment by johnvranos** [comment]
> @johnvranos You are compiling the kernel in a `/mnt/*` folder which can be problematic due to filename case-insensibility. Better extract and compile the kernel in a `/home/user/*` linux folder instead.

You are right!

I extracted to ~ (/home/john), and the compilation completed successfully, both in Ubuntu 21.10 (installed with Distrod), and Ubuntu-20.04 (installed from Microsoft Store).

Also the compilation was a lot faster, it took many hours previously.

So, I think the solution to this issue, is an additional instruction in the Build Instructions of https://github.com/microsoft/WSL2-Linux-Kernel, something like this:

> If you are building the kernel in Windows WSL, do not build it in any /mnt subdirectory.

**Comment by drokka** [comment]
I had the same build error re xt_HL.o target. I noticed the corresponding code file was xt_hl.c. So the default build rule failed.
I edited the Makefile replacing target xt_HL.o by xt_hl.o. It got past that target but failed on another xt_<UPPERCASE>.o target.
Yes I am building on a /mnt directory due to limited C drive space. The file name case insensitivity may be impacting if the build is renaming source files? In any case this does seem like an avoidable build issue. I'll see if repeating gets me through the build...
